lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java

```diff
@@ -231,62 +239,59 @@ public void execute() throws IOException, RecommenderBuildException {
                     if (recommender != null) {
                         recommender.close();
                     }
-
                     recommender = lre.createRecommender(config);
                 }
-                //gets prediction score
-                Result predictionResult = recommender.getRatingPredictor().predict(r.getUserId(), r.getItemId());
+
+                 //get prediction score
                 double predict = Double.NaN;
+                Result predictionResult = recommender.getRatingPredictor().predict(r.getUserId(), r.getItemId());
+
+                //check result to avoid null exception
                 if (predictionResult != null) {
                     predict = predictionResult.getScore();
                 }
-                //calculates time averaged RMSE
+                /***calculate Time Averaged RMSE***/
+                double rmse = 0.0;
                 if (!Double.isNaN(predict)) {
                     double err = predict - r.getValue();
                     sse += err * err;
                     n++;
-                }
-                double rmse = 0.0;
-                if (n > 0) {
                     rmse = sqrt(sse / n);
                 }
 
-                //calculate recommendation rank
-                int rank = -1;
-                LongSet itemsInDao = new LongOpenHashSet(limitedDao.getItemIds().size());
-                itemsInDao.addAll(limitedDao.getItemIds());
-
-                if (itemsInDao.size() > 0) {
-                    //candidate =  r.getItemId() + random (items from dao - items by user, listsize-1);
-                    LongSet itemsByUser = new LongOpenHashSet(itemsInDao.size());
-
-                    if (limitedDao.getEventsForUser(r.getUserId()) != null) {
-                        itemsByUser.addAll(limitedDao.getEventsForUser(r.getUserId()).itemSet());
-
-                    }
-                    itemsInDao.removeAll(itemsByUser);
-                    LongSet candidates = new LongOpenHashSet(10);
+                /***calculate recommendation rank***/
+                // item rank
+                int rank = -1; //rank < 0 indicates item not found
+                /* set of all items in limited DAO */
+                LongSet itemsInDao = new LongOpenHashSet(limitedDao.getItemIds());
+                /* set of all items rated by user */
+                LongSet itemsByUser = new LongOpenHashSet();
+                /* set of candidates that includes current item +
+                   listsize-1 random values from (items from dao - items rated by user) */
+                LongSet candidates = new LongOpenHashSet();
+
+                //Check if events for users exists to avoid NULL exception
+                if (limitedDao.getEventsForUser(r.getUserId()) != null) {
+                    itemsByUser.addAll(limitedDao.getEventsForUser(r.getUserId()).itemSet());
```

Comment:
If you also add the rating's item ID to `itemsByUser`, then you can unconditionally add it later and the whole thing will be more reliable.